sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/python/GpuWindowInPandasExecBase.scala

```diff
@@ -78,96 +77,265 @@ abstract class GpuWindowInPandasExecMetaBase(
+  private val queueBatches: mutable.Queue[(SpillableColumnarBatch, Seq[Int])] =
+    mutable.Queue.empty
 
   TaskContext.get().addTaskCompletionListener[Unit]{ _ =>
-    groupBatches.foreach(_.close())
-    groupBatches.clear()
+    queueBatches.foreach(_._1.close())
+    queueBatches.clear()
   }
 
-  override def hasNext(): Boolean = groupBatches.nonEmpty || wrapped.hasNext
+  override def hasNext(): Boolean = queueBatches.nonEmpty || wrapped.hasNext
 
-  override def next(): ColumnarBatch = {
-    if (groupBatches.nonEmpty) {
-      groupBatches.dequeue().getColumnarBatch()
+  // Return both the original batch and the additional info batch.
+  // The info batch will be used by the Python side.
+  override def next(): (ColumnarBatch, ColumnarBatch) = {
+    val (nextBatch, gpLens) = if (queueBatches.nonEmpty) {
+      val (sBatch, lens) = queueBatches.dequeue()
+      (sBatch.getColumnarBatch(), lens)
     } else {
       val batch = wrapped.next()
       inputBatches += 1
       inputRows += batch.numRows()
-      if (batch.numRows() > 0 && batch.numCols() > 0 && partitionSpec.nonEmpty) {
-        val partitionIndices = partitionSpec.indices
-        // 1) Calculate the split indices via cudf.Table.groupBy and aggregation Count.
-        //   a) Compute the count number for each group in a batch, including null values.
-        //   b) Restore the original order (Ascending with NullsFirst) defined in the plan,
-        //      since cudf.Table.groupBy will break the original order.
-        //   c) The 'count' column can be used to calculate the indices for split.
-        val cntTable = withResource(GpuProjectExec.project(batch, partitionSpec)) { projected =>
-          withResource(GpuColumnVector.from(projected)) { table =>
-            table
-              .groupBy(partitionIndices:_*)
-              .aggregate(Aggregation.count(true).onColumn(0))
-          }
+      val groupLens = genGroupInfoFromBatch(batch)
+      if (isSplitGroups && groupLens.size > 1) {
+        // Split the groups when being required and there are more than one group.
+        val groupBatchesAndSizes = splitBatchPerGroups(batch, groupLens)
+          .zip(groupLens.map(Seq(_)))
+        val tails = groupBatchesAndSizes.tail.map {
+          case (sb, gpInfo) =>
+            (SpillableColumnarBatch(sb, SpillPriorities.ACTIVE_ON_DECK_PRIORITY, spillCallback),
+              gpInfo)
         }
-        val orderedTable = withResource(cntTable) { table =>
-          table.orderBy(partitionIndices.map(id => Table.asc(id, true)): _*)
+        // Return the first one and enqueue others
+        queueBatches.enqueue(tails: _*)
+        groupBatchesAndSizes.head
+      } else {
+        (batch, groupLens)
+      }
+    }
+    (nextBatch, buildInfoBatchForPython(nextBatch, gpLens))
+  }
+
+  private[this] def splitBatchPerGroups(
+      batch: ColumnarBatch,
+      gpLens: Seq[Int]): Seq[ColumnarBatch] = {
+    val splitIndices = gpLens.dropRight(1).scan(0)(_+_).tail
```

Comment:
A comment here would be really helpful. It is not at all obvious why we need to remove the last element from the input sequence, do a running sum on the values, and then drop the first element from the list of indexes.

I think I understand why, but defining exactly what `gpLens` is meant to hold would make it much clearer what is happening.